src/Utils.cpp

```diff
@@ -499,6 +499,82 @@ CString CUtils::FormatTime(time_t t, const CString& sFormat,
     return s;
 }
 
+/* Supports additional format specifiers for formatting sub-second values:
+ *
+ * - %L - millisecond, 3 digits (ruby extension)
+ * - %N - sub-second fraction (ruby extension)
+ *   - %3N - millisecond
+ *   - %6N - microsecond
+ *   - %9N - nanosecond (default if no digit specified)
+ *
+ * However, note that timeval only supports microsecond precision
+ * (thus, formatting with higher-than-microsecond precision will
+ * always result in trailing zeroes), and IRC server-time is specified
+ * in millisecond precision (thus formatting received timestamps with
+ * higher-than-millisecond precision will always result in trailing
+ * zeroes).
+ */
+CString CUtils::FormatTime(timeval tv, const CString& sFormat,
+                           const CString& sTimezone) {
+    // Parse additional format specifiers before passing them to
+    // strftime, since the way strftime treats unknown format
+    // specifiers is undefined.
+    CString sFormat2;
+
+    // Make sure %% is parsed correctly, i.e. %%L is passed through to
+    // strftime to become %L, and not 123.
+    bool bInFormat = false;
+    int iDigits;
+    CString::size_type pLastCopied = 0, pFormatStart;
+
+    for (CString::size_type i = 0; i < sFormat.length(); i++) {
```

Comment:
Basically fine by me, but why not do a loop based on `.find('%')`? I'm not totally sure, but that might be easier/simpler.

Reply:
I don't see how that would be easier or simpler considering the constraints (see comments).